### PR TITLE
Feature/web monitor log parsing

### PR DESCRIPTION
**Pull request by icherkau**
Overview:
- Refactored and revamped Generate_combined_graph.ipynb into refreshingGraph.ipynb
- Added a pop-up interactive view in singleShotInteractiveGraph.py

PR Review Checklist

- [x] Check merging to correct branch
- [x] Confirm latest develop branch has been merged into PR
- [x] Confirm any merge conflicts are resolved
- [x] Confirm you have written (using AI) code logic and hardware tests and drafted clear test results so less human final review needed. Your PR should say where the test document is with results so final review can examine.
- [x] Run new branch code and confirm works when attached to real hardware, document results in a test document for the final reviewer
    - It is in this document: https://docs.google.com/document/d/1HkbC7HqXeKXyMz-Wmsc2SCqoWQ2bClf5BJ_eqAiOwZA/edit?tab=t.0
- [x] In VSCode with Codex/Claude extension review your own branch with multiple prompts similar to:
    - [x] "I want your assessment if this code is good to go. Tell me if the code is ready to merge into the develop branch. Is there anything critical that needs to change before merging? Will it break develop in any way? What do the changes in the new code do and are they good changes? Point out any critical errors, in particular logic errors or anything that might break the dashboard. Make sure you do a diff of this branch relative to develop."
- [x] Run @codex review from PR
- [x] Look through questions asked in the conversation and resolve any code questions (highlighting from files changed as needed) or comment why you're not addressing a question
- [x] Review entire "Files Changed" manually aided by AI:
    - [x] Confirm that no unnecessary development comments were committed
    - [x] Confirm only changes within the scope of this PR were committed
    - [x] Confirm no non-functional unnecessary changes were committed (e.g. needlessly adding extra space or something like that that just pollutes the PR review)
    - [x] Confirm new functional code makes sense
    - [x] Confirm hardware test
- [x] Let the final reviewer know ready for final review

### CODE REVIEW

**Comment by bwalkerMIR** [comment]
@codex review

**Comment by bwalkerMIR** [comment]
@icherkau are you still working on this or is it ready for review? I haven't done a full review yet, but here are some initial things from scanning around:
- your gitignore should be cleaned up a bit. For example excluding a specific "7-7-25 backup" is maybe a little odd, since other developers don't know what that is. Also you could just ignore the entire Data Samples folder and not the very specific "2025-05-29 file". 
- I don't understand why there are bunch of empty folders "Tera Term xxx" and "902b Logs" and "Misc Tera Term Logs"
- Also why is a PMON log commited to the repo?
- The "webMonitor_log" shouldn't be committed to the repo.
- It seems like there is a lot of extra stuff with log files and empty folders. Maybe you should just have an example data folder for testing with a representative set of data you do push to github but it's then clear what it is in a confined folder?
- The PR has +81,875 new lines of code. This is a bit large.. lol but is due to the webMonitor_log.txt. We can keep it if it would help I just think that stuff should be in a clear example data folder.
- requirements.txt should ideally lock to specific version so each devecloper doesn't download a different version of the dependency stack.

File: singleShotInteractiveGraph.py [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+
+    legacy_graph_dataframes = {
+        '20kV PSU voltage':   hv20kv_df,
+        '20kV PSU current':   hv20kv_df,
+        '3kV PSU voltage':    hv3kv_df,
+        '3kV PSU current':    hv3kv_df,
+        '1kV PSU voltage': hvPos1kv_df,
+        '1kV PSU current': hvPos1kv_df,
+        'Neg1kV PSU voltage': hvNeg1kv_df,
+        'Neg1kV PSU current': hvNeg1kv_df,
```
[the remark this answers is not in the document]
Comment (icherkau) [reply singleShotInteractiveGraph.py]: I've addressed this in commit [f0fe08a](https://github.com/uw-loci/cbmark_logger/pull/3/commits/f0fe08a60b87608445d3396d6c71a1f81266def5)

File: singleShotInteractiveGraph.py [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+        "unit": "mA",
+        "enabled": False
+    },
+        'pos1kV PSU voltage': {
+        "lines": ['hvActualVoltNeg1kv', 'hvSetVoltNeg1kv'],
+        "unit": "V",
+        "enabled": False
+    },
+    'neg1kV PSU current': {
+        "lines": ['hvCurrentNeg1kv'],
```
[the remark this answers is not in the document]
Comment (icherkau) [reply singleShotInteractiveGraph.py]: I updated I've addressed this in commit [f0fe08a](https://github.com/uw-loci/cbmark_logger/pull/3/commits/f0fe08a60b87608445d3396d6c71a1f81266def5)

File: singleShotInteractiveGraph.py [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+    teraTerm_log_file3kv = _most_recent_file(teraTerm_log_path3kv)
+    teraTerm_log_filePos1kv = _most_recent_file(teraTerm_log_pathPos1kv)
+    teraTerm_log_fileNeg1kv = _most_recent_file(teraTerm_log_pathNeg1kv)
+    teraTerm_log_file902b = _most_recent_file(teraTerm_log_path902b)
+    webMonitorFile = _most_recent_file(webMonitor_path)
+    dashboard_log_file = _most_recent_file(dashboard_log_path)
+
+
+    # Extract data from web monitor file and break up columns into different graphs
+    webMonitor_df = getDataFromWebMonitorFile(webMonitorFile)
```
[the remark this answers is not in the document]
Comment (icherkau) [reply singleShotInteractiveGraph.py]: I've addressed this in commit [32cbd80](https://github.com/uw-loci/cbmark_logger/pull/3/commits/32cbd80d02e30c87b501d1bb5b1b3b84bf8babc8)

**Comment by icherkau** [comment]
> @icherkau are you still working on this or is it ready for review? I haven't done a full review yet, but here are some initial things from scanning around:
> 
>     * your gitignore should be cleaned up a bit. For example excluding a specific "7-7-25 backup" is maybe a little odd, since other developers don't know what that is. Also you could just ignore the entire Data Samples folder and not the very specific "2025-05-29 file".
> 
>     * I don't understand why there are bunch of empty folders "Tera Term xxx" and "902b Logs" and "Misc Tera Term Logs"
> 
>     * Also why is a PMON log commited to the repo?
> 
>     * The "webMonitor_log" shouldn't be committed to the repo.
> 
>     * It seems like there is a lot of extra stuff with log files and empty folders. Maybe you should just have an example data folder for testing with a representative set of data you do push to github but it's then clear what it is in a confined folder?
> 
>     * The PR has +81,875 new lines of code. This is a bit large.. lol but is due to the webMonitor_log.txt. We can keep it if it would help I just think that stuff should be in a clear example data folder.
> 
>     * requirements.txt should ideally lock to specific version so each devecloper doesn't download a different version of the dependency stack.

I've addressed the folder organization in commit [d3823a5](https://github.com/uw-loci/cbmark_logger/pull/3/commits/d3823a52c399dfe51eae3c4ee8a37abde05db37d), and I've addressed the requirements.txt versions in commit [83a7779](https://github.com/uw-loci/cbmark_logger/pull/3/commits/83a77793664eb0057c607872299c4d4a8bb48d0b). I will be leaving the webMonitor file in so that we have a useful data WM sample, but I've relocated it to make it more clear that it's a data sample.

**Comment by bwalkerMIR** [comment]
@icherkau the structure looks much better, but there are still what looks from the outside like a lot of unnecessary files. Do all of these files really need to be committed to the repo for everyone? For example, archive has a ton of stuff. In github we generally don't have archive folders, because everything in the repo is either active worth keeping or is clearly labeled test or example data.

I would expect all example data to be in "Example data samples", but there are a lot of files in Archive.

Is there also example data in Tera Term Logs? This is in the gitignore, but then it has files in it. Are those files important? If those files are important, what about the empty blank file? Why not just use a gitkeep?

Is the blank.txt needed in Example Data? 

Would the Archive data be better to just refer in the readme to a gdrive location that has records of key experiment data from the past and future experiments?

**Comment by bwalkerMIR** [comment]
@codex review

File: static-interactive-graph.py [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+
+    # Used to only enable subplot if it has data
+    keepSubplotEnabled = False 
+
+    for subplot in graph_settings :
+        keepSubplotEnabled = False 
+
+        if(graph_settings[subplot]["enabled"]) :
+            for line in graph_settings[subplot]["lines"] :
+                if not(webMonitor_df[line].empty) :
```
[the remark this answers is not in the document]
Comment (icherkau) [reply static-interactive-graph.py]: Handled with commit [e16834a](https://github.com/uw-loci/cbmark_logger/pull/3/commits/e16834aa5f5a7511e9a7a94924553c54f065d90a)

File: static-interactive-graph.py [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+    plt.show()
+
+def wmCacheInit() :
+    # Read previous n webMonitor files (if available)
+    files = glob.glob(file_paths['webMonitor file path'])
+    if files:
+        maxIndex = min(int_settings['Number of previous WM Files to read'], len(files))
+        for i in range(0, maxIndex) :
+            files = sorted(files, key=os.path.getctime, reverse=True)
+            getDataFromWebMonitorFile(files[i])
```
[the remark this answers is not in the document]
Comment (icherkau) [reply static-interactive-graph.py:538]: Handled with commit [e16834a](https://github.com/uw-loci/cbmark_logger/pull/3/commits/e16834aa5f5a7511e9a7a94924553c54f065d90a)

**Comment by icherkau** [comment]
> @icherkau the structure looks much better, but there are still what looks from the outside like a lot of unnecessary files. Do all of these files really need to be committed to the repo for everyone? For example, archive has a ton of stuff. In github we generally don't have archive folders, because everything in the repo is either active worth keeping or is clearly labeled test or example data.
> 
> I would expect all example data to be in "Example data samples", but there are a lot of files in Archive.
> 
> Is there also example data in Tera Term Logs? This is in the gitignore, but then it has files in it. Are those files important? If those files are important, what about the empty blank file? Why not just use a gitkeep?
> 
> Is the blank.txt needed in Example Data?
> 
> Would the Archive data be better to just refer in the readme to a gdrive location that has records of key experiment data from the past and future experiments?

I wouldn't mind the Archive being put in Google Drive instead. Let me know if you'd like that to happen. It certainly would make this PR a lot bigger, though. 

I've moved the example data from the Tera Term folder, but I'm definitely keeping blank.txt, because it's useful for testing files with no data and it doesn't take up space. It can also be used to disable a subplot, but that's not super relevant anymore with the control panel buttoms

**Comment by bwalkerMIR** [comment]
@icherkau yeah if Archive isn't actively used let's put that in gdrive. I like github repos to be very clean for developers so it's easy to understand what everything is. For the example data does all of that need to be there? It just looks like there is maybe a lot of redundancy there.

**Comment by bwalkerMIR** [comment]
@codex review

File: static-interactive-graph.py [hunk 6]
```diff
@@ -0,0 +1,644 @@
...
+        filename -> str
+    
+    @return:
+        2D list of time and HV current -> list
+    '''
+    
+    # Create an empty list to store the extracted data before converting it to a DataFrame
+    data = []                          
+    # Regex pattern to parse lines in the Tera Term log file for timestamps and HV PSU readings (voltage set point, voltage actual, and current)
+    regex_pattern = re.compile(r'\[\d{4}-\d{2}-\d{2} (.*?)\] Set: -?(\d{1,4}) V,  HV: -?(\d{1,4}) V,  I: -?(\d{1,2}\.\d{2,3}) mA', re.I)
```
[the remark this answers is not in the document]
Comment (bwalkerMIR) [reply static-interactive-graph.py]: @icherkau what do you think of this one?

**Comment by bwalkerMIR** [comment]
Also @icherkau here is a codex 5.5 xhigh review of this PR. Can you respond to each whether something that should be fixed now, or made an issue for a future focused PR to fix, or is not worth fixing at this time, or is not really an error?

**Findings**

1. Severity: `high`  
File/function: [static-interactive-graph.py](C:/Users/brand/Code/cbmark_logger/static-interactive-graph.py:279) and [live-graph.ipynb](C:/Users/brand/Code/cbmark_logger/live-graph.ipynb:310), `getDataFromWebMonitorFile`  
Problem: WebMonitor rows are deduped only by timestamp. The sample log has 80,456 rows but only 3,601 unique timestamps, with up to 32 rows in the same second, so this drops most readings.  
Why it matters: Live graphs and CSV exports silently lose high-frequency data.  
Fix: Do not dedupe by timestamp alone. Preserve duplicate datetime indexes, or dedupe by a real row identity such as file path plus byte offset/line number. The tail offset already prevents normal duplicate reads.

2. Severity: `medium`  
File/function: [static-interactive-graph.py](C:/Users/brand/Code/cbmark_logger/static-interactive-graph.py:446) and [live-graph.ipynb](C:/Users/brand/Code/cbmark_logger/live-graph.ipynb:477), `getHVData`  
Problem: The HV regex allows negative values but puts the `-?` outside the capture groups, so negative set voltage, actual voltage, and current are parsed as positive.  
Why it matters: The `-1kV` plots and any negative readings in the included sample/archive logs are numerically wrong.  
Fix: Capture the sign: `(-?\d{1,4})` and `(-?\d{1,2}\.\d{2,3})`.

3. Severity: `medium`  
File/function: [static-interactive-graph.py](C:/Users/brand/Code/cbmark_logger/static-interactive-graph.py:522) and [live-graph.ipynb](C:/Users/brand/Code/cbmark_logger/live-graph.ipynb:548), `getNumPlots`  
Problem: The function overwrites each user setting’s `"enabled"` flag with whether data exists right now. A transient missing/empty file disables that subplot permanently until the user manually re-enables it.  
Why it matters: A live dashboard can stay blank even after logs appear or recover.  
Fix: Keep user intent separate from data availability. Return a local list/set of drawable subplots, or add a separate `has_data` flag without mutating `"enabled"`.

4. Severity: `medium`  
File/function: [Tera Term Macros/3kv HV monitor log.ttl](<C:/Users/brand/Code/cbmark_logger/Tera Term Macros/3kv HV monitor log.ttl:7>), [Tera Term Macros/HV monitor log.ttl](<C:/Users/brand/Code/cbmark_logger/Tera Term Macros/HV monitor log.ttl:7>), [Tera Term Macros/Recurring 902b pressure query.ttl](<C:/Users/brand/Code/cbmark_logger/Tera Term Macros/Recurring 902b pressure query.ttl:7>)  
Problem: The PR deletes the old log folders, but the Tera Term macros still write to those deleted paths.  
Why it matters: Macro logging will fail or require manual folder recreation after a clean clone.  
Fix: Either restore those folders and ignore rules, or update the macros to write into the new supported `Tera Term logs/` structure.

**Top 3 Risks**

1. WebMonitor data is massively downsampled unintentionally.
2. Negative HV readings are plotted/exported with the wrong sign.
3. Live graph settings can disable themselves after transient missing data.

**Suggested Tests**

Add parser tests for duplicate WebMonitor timestamps, negative HV values, and missing/empty log files followed by later valid data. Add a clean-clone smoke test that verifies documented Tera Term macro output paths exist.

**Overall Verdict**

`request changes`

**Comment by icherkau** [comment]
I have addressed the Tera Term Macro filepath issue with commit [d570bee](https://github.com/uw-loci/cbmark_logger/pull/3/commits/d570bee08926e3e62fe3edf107517b32f53eeb3d).

**Comment by icherkau** [comment]
I have addressed issues 2 and 3 with commits [30ac030](https://github.com/uw-loci/cbmark_logger/pull/3/commits/30ac0309fe0bc5aae43e6b43b15a48adc74d3fbe) and [dfdf4f1](https://github.com/uw-loci/cbmark_logger/pull/3/commits/dfdf4f1bbad2b3d476d90f25576a6a1cd821a885)

**Comment by icherkau** [comment]
> Also @icherkau here is a codex 5.5 xhigh review of this PR. Can you respond to each whether something that should be fixed now, or made an issue for a future focused PR to fix, or is not worth fixing at this time, or is not really an error?
> 
> **Findings**
> 
>     1. Severity: `high`
>        File/function: static-interactive-graph.py and live-graph.ipynb, `getDataFromWebMonitorFile`
>        Problem: WebMonitor rows are deduped only by timestamp. The sample log has 80,456 rows but only 3,601 unique timestamps, with up to 32 rows in the same second, so this drops most readings.
>        Why it matters: Live graphs and CSV exports silently lose high-frequency data.
>        Fix: Do not dedupe by timestamp alone. Preserve duplicate datetime indexes, or dedupe by a real row identity such as file path plus byte offset/line number. The tail offset already prevents normal duplicate reads.
> 
>     2. Severity: `medium`
>        File/function: static-interactive-graph.py and live-graph.ipynb, `getHVData`
>        Problem: The HV regex allows negative values but puts the `-?` outside the capture groups, so negative set voltage, actual voltage, and current are parsed as positive.
>        Why it matters: The `-1kV` plots and any negative readings in the included sample/archive logs are numerically wrong.
>        Fix: Capture the sign: `(-?\d{1,4})` and `(-?\d{1,2}\.\d{2,3})`.
> 
>     3. Severity: `medium`
>        File/function: static-interactive-graph.py and live-graph.ipynb, `getNumPlots`
>        Problem: The function overwrites each user setting’s `"enabled"` flag with whether data exists right now. A transient missing/empty file disables that subplot permanently until the user manually re-enables it.
>        Why it matters: A live dashboard can stay blank even after logs appear or recover.
>        Fix: Keep user intent separate from data availability. Return a local list/set of drawable subplots, or add a separate `has_data` flag without mutating `"enabled"`.
> 
>     4. Severity: `medium`
>        File/function: Tera Term Macros/3kv HV monitor log.ttl, Tera Term Macros/HV monitor log.ttl, Tera Term Macros/Recurring 902b pressure query.ttl
>        Problem: The PR deletes the old log folders, but the Tera Term macros still write to those deleted paths.
>        Why it matters: Macro logging will fail or require manual folder recreation after a clean clone.
>        Fix: Either restore those folders and ignore rules, or update the macros to write into the new supported `Tera Term logs/` structure.
> 
> 
> **Top 3 Risks**
> 
>     1. WebMonitor data is massively downsampled unintentionally.
> 
>     2. Negative HV readings are plotted/exported with the wrong sign.
> 
>     3. Live graph settings can disable themselves after transient missing data.
> 
> 
> **Suggested Tests**
> 
> Add parser tests for duplicate WebMonitor timestamps, negative HV values, and missing/empty log files followed by later valid data. Add a clean-clone smoke test that verifies documented Tera Term macro output paths exist.
> 
> **Overall Verdict**
> 
> `request changes`

As mentioned above, I have fixed issues 2-4 from this analysis. Issue 1 is a result of missing data and will require a fix in the Dashboard, and cannot be resolved in CBMark_logger without faking data. Without this fix, CBMark_logger will only display one point in every timestamp (the last one) and miss some high-frequency data. However, the Dashboard has been updated on its develop branch to have a new WM log file format. As a result, this CBMark_logger version will not parse Dashboard develop branch log files and I cannot implement this fix. However, this branch should still be merged to save a version that is compatible with the older log file format currently on the Dashboard main branch. This version of CBMark_logger in combination with the Dashboard main version still works, but will not display all data points. I have created new issues for this, which will be resolved in a future branch. 

Note: CBMark_logger will currently display timestamps with miliseconds (which is what the dashboard should have been outputting), but there are no miliseconds to read.

**Comment by icherkau** [comment]
Reminder for Brandon: This PR is supposed to work with the Dashboard main branch, not the Dashboard develop branch

**Comment by icherkau** [comment]
@codex review

**Comment by icherkau** [comment]
@codex review

File: static-interactive-graph.py [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+
+    # Append new rows into the cache.
+    if _webmon_cache_df is None or _webmon_cache_df.empty:
+        _webmon_cache_df = new_df
+    else:
+        _webmon_cache_df = pd.concat([_webmon_cache_df, new_df])
+
+    # Keep index ordered and dedupe timestamps (last write wins).
+    _webmon_cache_df = _webmon_cache_df.sort_index()
+    _webmon_cache_df = _webmon_cache_df[~_webmon_cache_df.index.duplicated(keep="last")]
```
[the remark this answers is not in the document]
Comment (icherkau) [reply static-interactive-graph.py:303]: 👍 

Addressed with commit [3dc2506](https://github.com/uw-loci/cbmark_logger/pull/3/commits/3dc2506a691c4c0ca93e77749f6b539f0d925377)

**Comment by icherkau** [comment]
@codex review

File: static-interactive-graph.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (icherkau) [reply static-interactive-graph.py]: Yes, this should be addressed. I've fixed it in the most recent commit [426a960](https://github.com/uw-loci/cbmark_logger/pull/3/commits/426a96059cb027b6a6e0e554db793724be929ba1).

**Comment by icherkau** [comment]
> Also @icherkau here is a codex 5.5 xhigh review of this PR. Can you respond to each whether something that should be fixed now, or made an issue for a future focused PR to fix, or is not worth fixing at this time, or is not really an error?
> 
> **Findings**
> 
>     1. Severity: `high`
>        File/function: static-interactive-graph.py and live-graph.ipynb, `getDataFromWebMonitorFile`
>        Problem: WebMonitor rows are deduped only by timestamp. The sample log has 80,456 rows but only 3,601 unique timestamps, with up to 32 rows in the same second, so this drops most readings.
>        Why it matters: Live graphs and CSV exports silently lose high-frequency data.
>        Fix: Do not dedupe by timestamp alone. Preserve duplicate datetime indexes, or dedupe by a real row identity such as file path plus byte offset/line number. The tail offset already prevents normal duplicate reads.
> 
>     2. Severity: `medium`
>        File/function: static-interactive-graph.py and live-graph.ipynb, `getHVData`
>        Problem: The HV regex allows negative values but puts the `-?` outside the capture groups, so negative set voltage, actual voltage, and current are parsed as positive.
>        Why it matters: The `-1kV` plots and any negative readings in the included sample/archive logs are numerically wrong.
>        Fix: Capture the sign: `(-?\d{1,4})` and `(-?\d{1,2}\.\d{2,3})`.
> 
>     3. Severity: `medium`
>        File/function: static-interactive-graph.py and live-graph.ipynb, `getNumPlots`
>        Problem: The function overwrites each user setting’s `"enabled"` flag with whether data exists right now. A transient missing/empty file disables that subplot permanently until the user manually re-enables it.
>        Why it matters: A live dashboard can stay blank even after logs appear or recover.
>        Fix: Keep user intent separate from data availability. Return a local list/set of drawable subplots, or add a separate `has_data` flag without mutating `"enabled"`.
> 
>     4. Severity: `medium`
>        File/function: Tera Term Macros/3kv HV monitor log.ttl, Tera Term Macros/HV monitor log.ttl, Tera Term Macros/Recurring 902b pressure query.ttl
>        Problem: The PR deletes the old log folders, but the Tera Term macros still write to those deleted paths.
>        Why it matters: Macro logging will fail or require manual folder recreation after a clean clone.
>        Fix: Either restore those folders and ignore rules, or update the macros to write into the new supported `Tera Term logs/` structure.
> 
> 
> **Top 3 Risks**
> 
>     1. WebMonitor data is massively downsampled unintentionally.
> 
>     2. Negative HV readings are plotted/exported with the wrong sign.
> 
>     3. Live graph settings can disable themselves after transient missing data.
> 
> 
> **Suggested Tests**
> 
> Add parser tests for duplicate WebMonitor timestamps, negative HV values, and missing/empty log files followed by later valid data. Add a clean-clone smoke test that verifies documented Tera Term macro output paths exist.
> 
> **Overall Verdict**
> 
> `request changes`

Parts 2-4 have been addressed in previous commits. I made comments about them, but they seem to have disappeared. Part 1 has been addressed in the most recent branch, but I had to do a fair amount of restructuring, and I would like to not change that here because I don't want to have other uncaught side effects .

**Comment by bwalkerMIR** [comment]
@icherkau says good to review now

**Comment by bwalkerMIR** [comment]
@codex review

File: live-graph.ipynb [hunk 8]
```diff
@@ -0,0 +1,750 @@
...
+    "\n",
+    "    # Append new rows into the cache.\n",
+    "    if _webmon_cache_df is None or _webmon_cache_df.empty:\n",
+    "        _webmon_cache_df = new_df\n",
+    "    else:\n",
+    "        _webmon_cache_df = pd.concat([_webmon_cache_df, new_df])\n",
+    "\n",
+    "    # Keep index ordered and dedupe timestamps (last write wins).\n",
+    "    _webmon_cache_df = _webmon_cache_df.sort_index()\n",
+    "    _webmon_cache_df = _webmon_cache_df[~_webmon_cache_df.index.duplicated(keep=\"last\")]\n",
```
[the remark this answers is not in the document]
Comment (icherkau) [reply live-graph.ipynb:317]: I have resolved this in the most recent branch and chose not to address it in this PR because it would require too much repeated work and testing.

**Review by bwalkerMIR — APPROVED** [review]
This is a major overhaul of the plotter and drastically cuts down on unecessary files. Approved.
